### ISSUE TITLE
Syntax error with a top-level await and no import

### ISSUE THREAD

**Issue by regseb** [mozilla/addons-linter#4020]
### Describe the problem and steps to reproduce it:

Files with a _top-level await_ and no _import_: returns a syntax error.

```JavaScript
await browser.storage.local.get();
```

### What happened?

```
Validation Summary:

errors          1              
notices         0              
warnings        0              

ERRORS:

Code              Message        Description               File          Line   Column
JS_SYNTAX_ERROR   JavaScript     There is a JavaScript     testcase.js   1      7     
                  syntax error   syntax error in your                                
                   (Parsing as   code, which might be                                
                  script         related to some                                     
                  error:         experimental JavaScript                             
                  Unexpected     features that aren't an                             
                  token          official part of the                                
                  browser at     language specification                              
                  line: 1 and    and therefore not                                   
                  column: 7)     supported yet. The                                  
                                 validation cannot                                   
                                 continue on this file.
```

### What did you expect to happen?

No error reported.

### Anything else we should know?

I have no error with ESLint.

**Comment by willdurand** [comment]
You have to use a `.mjs` file I believe.

**Comment by regseb** [comment]
I found out why it works with ESLint: I set the [`sourceType`](https://eslint.org/docs/user-guide/configuring/language-options#specifying-parser-options) to `"module"`.

The convention is to use the _.js_ extension for JavaScript files in browsers. The _.mjs_ extension is used by Node.js. I think it isn't a good practice to adapt the file name only for testing purposes.

All the functionality of the _scripts_ is in the _modules_. So we could test all files with `sourceType` to `"module"`?

**Comment by willdurand** [comment]
`mjs` files define ES modules and those files are not node-specific. The linter already supports them: https://github.com/mozilla/addons-linter/blob/cccc4e3588d1d55a160b1736d2cdcad24e305836/src/scanners/javascript.js#L274

We might be able to improve our source type inference but there is nothing wrong with using `mjs` files.

**Comment by regseb** [comment]
_.mjs_ files are interpreted as ES Module only in Node (and addons-linter). In browsers, _.mjs_ files are interpreted as _.js_ files. With this manifest, we think that the _background.mjs_ file is loaded as an ES Module, but it's interpreted as a _script_.

```JSON
{
    "background": {
        "scripts": ["background.js"]
    }
}
```

Is there a counter-argument to analyze all the files with the `sourceType` to `"module"`?

**Comment by willdurand** [comment]
Ah, right. `background.scripts` isn't able to load ES modules as per https://bugzilla.mozilla.org/show_bug.cgi?id=1394303. The encouraged approach seems to be the use of `background.page`. I think `.mjs` would be allowed here (with or without `type="module"`, I haven't checked).

> Is there a counter-argument to analyze all the files with the `sourceType` to `"module"`?

Not that I am aware of. I don't know why our source type inference is based on the content (and on only a few AST nodes). Generally speaking, inverting a default value or behavior isn't generally without side effects so we'd need to be careful.

**Comment by regseb** [comment]
I tested with a _background.mjs_ file containing an `import`. I loaded with a `<script>` without `type` attribut in the _background.page_ and I have the error:

> Uncaught SyntaxError: import declarations may only appear at top level of a module

**Comment by regseb** [comment]
I made an analysis (with three files) of the current solution of source type detection with `import` / `export`; and of my proposal to always analyze source with the [`sourceType`](https://eslint.org/docs/user-guide/configuring/language-options#specifying-parser-options) to `"module"`.

- then.js:

  ```JavaScript
  Promise.resolve(true).then((t) => console.log(t));
  ```

- export.js:

  ```JavaScript
  const t = await Promise.resolve(true);
  console.log(t);

  export default "foo";
  ```

- await.js:

  ```JavaScript
  const t = await Promise.resolve(true);
  console.log(t);
  ```

<table>
  <tr>
    <th rowspan="2">Files</th>
    <th rowspan="2">
      Load in browser with
      <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-type">type</a>:
    </th>
    <th rowspan="2">Browser</th>
    <th colspan="2">addons-linter with</th>
  </tr>
  <tr>
    <th>
      <a href="https://github.com/mozilla/addons-linter/blob/cccc4e3588d1d55a160b1736d2cdcad24e305836/src/scanners/javascript.js#L212"><code>_getSourceType()</code></a>
    </th>
    <th>
      <code>{sourceType: "module"}</code>
    </th>
  </tr>
  <tr>
    <td rowspan="2">then.js</td>
    <td><code>text/javascript</code></td>
    <td><code>true</code></td>
    <td>:heavy_check_mark:</td>
    <td>:heavy_check_mark:</td>
  </tr>
  <tr>
    <td><code>module</code></td>
    <td><code>true</code></td>
    <td>:heavy_check_mark:</td>
    <td>:heavy_check_mark:</td>
  </tr>
  <tr>
    <td rowspan="2">export.js</td>
    <td><code>text/javascript</code></td>
    <td>
      <em>Uncaught SyntaxError: await is only valid in async functions, async
      generators and modules</em>
    </td>
    <td>:x: (false negative)</td>
    <td>:x: (false negative)</td>
  </tr>
  <tr>
    <td><code>module</code></td>
    <td><code>true</code></td>
    <td>:heavy_check_mark:</td>
    <td>:heavy_check_mark:</td>
  </tr>
  <tr>
    <td rowspan="2">await.js</td>
    <td><code>text/javascript</code></td>
    <td>
      <em>Uncaught SyntaxError: await is only valid in async functions, async
      generators and modules</em>
    </td>
    <td>:x: (false negative)</td>
    <td>:x: (false negative)</td>
  </tr>
  <tr>
    <td><code>module</code></td>
    <td><code>true</code></td>
    <td>
      :x: (false positive)<br />
      <em>JavaScript syntax error (Parsing as script error: Unexpected token
      Promise at line: 1 and column: 17)</em>
    </td>
    <td>:heavy_check_mark:</td>
  </tr>
</table>

Caption:

- :heavy_check_mark:: addons-linter has the same behavior as the browser.
- :x: (false negative): addons-linter doesn't report the error that will appear in the browser.
- :x: (false positive): addons-linter reports a false error that will not appear in the browser.

With `{sourceType: "module"}`: there are always the false negatives :disappointed:, but we remove a false positive :sunglasses:.

**Comment by Rob--W** [comment]
We're open to exploring whether we can improve the detection by detecting the use of top-level await.

Are you interested in creating a patch with a fix + unit tests @regseb? The relevant code is at https://github.com/mozilla/addons-linter/blob/cccc4e3588d1d55a160b1736d2cdcad24e305836/src/scanners/javascript.js#L212-L247.

**Comment by regseb** [comment]
```Javascript
const eval = true;                                                               
console.log(eval);
```

The above code is valid in a _script_. In a _module_, Firefox returns an _Uncaught SyntaxError: 'eval' can't be defined or assigned to in strict mode code_.

There is a problem when analyzing all the files with the [`sourceType`](https://eslint.org/docs/user-guide/configuring/language-options#specifying-parser-options) to `"module"` (https://github.com/regseb/addons-linter/commit/4c1bfff158984cb2b2108c4bc6c3e08d90bda792), addons-linter will return a false positive for _scripts_ redeclaring `eval`. 🙁

**Comment by Rob--W** [comment]
> ```Javascript
> const eval = true;                                                               
> console.log(eval);
> ```
> 
> The above code is valid in a _script_. In a _module_, Firefox returns an _Uncaught SyntaxError: 'eval' can't be defined or assigned to in strict mode code_.
> 
> There is a problem when analyzing all the files with the [`sourceType`](https://eslint.org/docs/user-guide/configuring/language-options#specifying-parser-options) to `"module"` (https://github.com/regseb/addons-linter/commit/4c1bfff158984cb2b2108c4bc6c3e08d90bda792), addons-linter will return a false positive for _scripts_ redeclaring `eval`. 🙁
> 

This doesn't necessarily sound like a part of this report here. Rejecting a redeclaration of `eval` sounds reasonable to me.

@regseb Have you seen my last reply? Given your interest in this feature, I asked whether you were interested in contributing a patch.

**Comment by regseb** [comment]
@Rob--W I created the pull request #4120 to improve `sourceType` detection.

**Comment by regseb** [comment]
I created a pull request https://github.com/mozilla/addons-linter/pull/4795 to detect files with an top-level `await` and parse them with `module` type.

**Comment by willdurand** [comment]
FWIW I think we can now specify `"type": "module"` in the manifest,see: https://bugzilla.mozilla.org/show_bug.cgi?id=1811443 and https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Background_scripts#specify_the_background_scripts

**Comment by ioanarusiczki** [comment]
@willdurand 

With linter 6.2.0 I don't get the error [above](https://github.com/mozilla/addons-linter/issues/4020#issue-1043712102)

This is before and after the fix: 

![before and after](https://user-images.githubusercontent.com/33448286/235142596-dad83e40-17aa-45e4-827f-76a22e75ae37.png)

**Comment by regseb** [comment]
Detection works with [`await Promise.resolve();`](https://astexplorer.net/#/gist/50a751246e1ebe68461e28b46b6f1e06/3b25a9ec2c5e0546e0b90f49645953ba5703e55d), but not with [`const foo = await Promise.resolve();`](https://astexplorer.net/#/gist/b1572e78f7941d17aef29d5f1e73da4d/71666265ad0b8a88e68b0dda7b744056b64d9daa).

**Comment by ioanarusiczki** [comment]
@regseb

I tried on linter 6.2.0 with `const foo = await Promise.resolve();` -> got the error [mentioned](https://github.com/mozilla/addons-linter/issues/4020#issue-1043712102) 

Then with 6.3.0 -> addon passed validation without problems.
